client/src/main/java/org/eclipse/hono/client/CommandClient.java

```diff
@@ -0,0 +1,53 @@
+package org.eclipse.hono.client;
+
+import io.vertx.core.Future;
+
+/**
+ * A client for accessing Hono's Command and Control API.
+ * <p>
+ * An instance of this interface is always scoped to a specific tenant.
+ * </p>
+ */
+public interface CommandClient extends RequestResponseClient {
+
+    /**
+     * Sends a command to a device and expects a response.
+     * <p>
+     * A device needs to be (successfully) registered before a client can upload
+     * any data for it. The device also needs to be connected for a successful delivery.
+     *
+     * @param command The command name.
+     * @param data The command data to send to the device.
+     * @return A future indicating the result of the operation.
+     *         <p>
+     *         The future will succeed if a response with status 201 has been received from the device/adapter.
+     *         <p>
+     *         Otherwise, the future will fail with a {@link ServiceInvocationException} containing
+     *         the (error) status code returned by the service.
+     * @throws NullPointerException if device ID is {@code null}.
+     * @see RequestResponseClient#setRequestTimeout(long)
+     */
+    Future<byte[]> commandWithResponse(String command, byte[] data);
```

Comment:
Could we think of using `ByteBuffer`? That could prevent a few copies, and we could still add `byte[]` using default methods for simplicity.

Reply:
or vert.x `Buffer`?

Reply:
Since internally we use the vert.x Buffer and the Buffer implementation uses netty ByteBuf internally (and not ByteBuffer), the use of Buffer seems better. But on the other hand the user need to use the Vert.x objects (or at least the netty ByteBuf) to prevent copies.